src/packer.ts

```diff
@@ -0,0 +1,195 @@
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    fixed: Map<string, any>,
+
+    // custom function to do additional unpacking logic on the unpacked map
+    // TODO: sort out types..
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    process: (unpacked: Map<string, any>) => Map<string, any>
+}
+
+// For a given name and associated shape, do some validation, make shape type consistent and return number of values
+// after expansion
+const preparePackArrayForShape = (name: string, shape: number |  number[] | null) => {
+    // Scalars can be defined within the array option, if that allows user's preferred ordering
+    const scalar = shape === null || (Array.isArray(shape) && shape.length === 0);
+    if (scalar) {
+        return { names: [name], shape: [0], n: 1 };
+    }
+    const arrayShape: Array<number> = typeof shape === "number" ? [shape] : shape;
+    const nonInteger = arrayShape.find((v) => !Number.isInteger(v))
+    if (nonInteger) {
+        throw Error("All dimension values in 'array' values must be integers, but this is not the case for " +
+                    `${name}, whose value is ${JSON.stringify(shape)}`);
+    }
+    const lessThanZero = arrayShape.find((v) => v <= 0);
+    if (lessThanZero) {
+        throw Error("All dimension values in 'array' values must be at least 1, but this is not the case for " +
+                    `${name}, whose value is ${JSON.stringify(shape)}`);
+    }
+
+    // total number of values in shape
+    const n = arrayShape.reduce((prev, current) => prev * current, 1);
+
+    return {
+        shape: arrayShape,
+        n
+    }
+}
+
+// interface describing the range of indexes for a named value group in the underlying array
+interface IndexValues {
+    start: number,
+    length: number
+}
+
+export class Packer {
+    private options: Partial<PackerOptions>;
+    private len: number; // Total number of values
+    private idx: Record<string, IndexValues>;
+    private shape: Record<string, number[]>; // shape of each named value group i.e. size of each dimension
+
+    constructor(options: Partial<PackerOptions>) {
+        this.options = options;
+
+        this.idx = {};
+        this.shape = {};
+
+        const { scalar, fixed } = options;
+        const arrayOptions = options.array;
+
+        const allNames = [...(scalar || []), ...(arrayOptions?.keys() || []), ...(fixed?.keys() || [])];
+        // Check for duplicate names
+        const dup = allNames.find((name, i) => allNames.lastIndexOf(name) !== i)
+        if (dup) {
+            // TODO: make a throw error util
+            throw Error(`Names must be distinct between 'scalar', 'array' and 'fixed'. ${dup} appears in more ` +
+                        "than one place.");
+        }
+
+        if (scalar) {
+            Array.from(scalar).forEach((name, i) => {
+                // for each scalar, shape gets an array with a single integer in it, set to 0 (indicating no dimensions)
```

Comment:
why not use `[]` as the empty list? This has the nice property that `prod([]) == 1` so the product over dimensions works without special casing.

If you're basing this off the R version, `integer(0)` is really an integer vector of length 0, not an integer vector containing 0